### PR TITLE
fix(metrics): use Welford's algorithm in PearsonCorrelation for numerical stability

### LINKED ISSUE #3662: [Bug] Numerical instability in `PearsonCorrelation` due to naive variance formula
The current implementation of `PearsonCorrelation` in `ignite.metrics.regression` uses the **naive "sum of squares" algorithm** to compute variance and covariance:

\[
Var(X) = E[X^2] - (E[X])^2
\]

Although mathematically valid, this approach is **numerically unstable** when the input values have **large magnitudes relative to their variance**.

This leads to **catastrophic cancellation**, where two very large numbers (`E[X^2]` and `(E[X])^2`) are subtracted, causing loss of precision in **float32**. As a result, the metric can produce **incorrect results**, such as returning **0.0** when the true correlation is **≈ 0.707**.

---

## Steps to Reproduce

```python
import torch
from ignite.metrics.regression import PearsonCorrelation
import numpy as np

# Magnitude that triggers precision loss in float32
offset = 1e8 

y_true = torch.tensor([1.0, 2.0, 3.0, 4.0, 5.0]) + offset
y_pred = torch.tensor([1.1, 2.1, 3.1, 4.1, 5.1]) + offset

metric = PearsonCorrelation()
metric.update((y_pred, y_true))
ignite_res = metric.compute()

# Ground truth using PyTorch's stable corrcoef
combined = torch.stack([y_pred, y_true])
torch_res = torch.corrcoef(combined)[0, 1].item()

print(f"Offset used: {offset}")
print(f"Ignite Result: {ignite_res}")
print(f"Torch Ground Truth: {torch_res}")
```
---
## actual output

```python
Offset used: 100000000.0
Ignite Result: 0.0
Torch Ground Truth: 0.7071067690849304
```
---
## Expected Behavior

The **Pearson correlation metric** should be **invariant to constant offsets** in the input data.

Therefore, the result should be consistent with **numerically stable implementations** such as:

- **`torch.corrcoef`**
- **`scipy.stats.pearsonr`**

###  Expected Result

```text
≈ 0.7071067
```
---
##  Proposed Fix

To improve **numerical stability**, the implementation can be updated as follows:

### Use a Numerically Stable Online Algorithm
Implement **Welford’s Online Algorithm** (or a similar **one-pass algorithm**) to compute:

- **Mean**
- **Variance**
- **Covariance**

incrementally.

---

### Use Higher Precision Accumulators
Maintain internal accumulators using **`torch.float64`** to prevent **precision loss during batch updates**.

---

### Ensure Compatibility with Ignite Metrics
The updated implementation should still support **batch-wise streaming updates** used by **Ignite metrics**.

---

### Benefit
These changes will prevent **catastrophic cancellation** and ensure **correct results** even when the data has:

- **Large offsets**
- **High magnitude values**
---
Hi @vfdev-5, I would like to work on a PR to resolve this issue if this looks good to you!

### PR DESCRIPTION
**Pull request by joemunene-by**
## Summary

Fixes #3662

- Replace the naive sum-of-squares variance formula (`E[x²] - E[x]²`) with [Welford's online algorithm](https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford%27s_online_algorithm) for numerically stable computation
- Internal accumulators use float64 to further guard against precision loss
- Distributed sync uses parallel Welford merge via `idist.all_gather` instead of sum-based `@sync_all_reduce`

### The problem

The naive formula suffers from **catastrophic cancellation** in float32 when input values have large magnitudes relative to their variance. For example, with `mean=1e6` and `std=1`:

| Method | Result | Error vs scipy |
|--------|--------|----------------|
| Old (naive float32) | 0.8944 | **10.0%** |
| New (Welford float64) | 0.9947 | **0.003%** |

### Changes

- **`pearson_correlation.py`**: Replaced sum-based state (`_sum_of_y_pred_squares`, etc.) with Welford state (`_mean_x`, `_mean_y`, `_m2_x`, `_m2_y`, `_cxy`). Batch updates use the parallel variant of Welford's algorithm. Distributed merge gathers per-process states and merges them pairwise.
- **`test_pearson_correlation.py`**: Updated accumulator attribute names in tests. Added `test_numerical_stability` that validates correctness with large-mean, small-variance data fed in small batches.

## Test plan

- [x] All existing tests pass (`test_zero_sample`, `test_wrong_input_shapes`, `test_degenerated_sample`, `test_pearson_correlation`, `test_integration_pearson_correlation`, `test_accumulator_detached`)
- [x] New `test_numerical_stability` passes — verifies the fix against the exact scenario described in #3662
- [ ] Distributed tests (`TestDistributed`) — updated attribute names, needs CI with distributed backends to verify

### CODE REVIEW

**Comment by aaishwarymishra** [comment]
I think variables names should be changed, as while I am reviewing I have no idea which variables is for what as there name is so confusing.

**Comment by aaishwarymishra** [comment]
I think we should wait for now as there is one more pr for issue :)

**Comment by joemunene-by** [comment]
Thanks for the review! You're right that the accumulator names are opaque without context — I kept `_m2_x` / `_cxy` because they mirror the Welford / Chan et al. literature, but added a block comment mapping each attribute to its role, plus inline notes on the batch-merge and distributed-merge formulas (why a plain sum across ranks is wrong, what the `n_a·n_b/n_ab · Δ²` correction term represents). Pushed in df1fde02.

Happy to wait for maintainer direction on which of this PR and #3740 to land — the two approaches are complementary (float64 alone vs. algorithmic stability via Welford). Noted in #3662 that the shared-utility refactor across `PearsonCorrelation`, `R2Score`, and `FID` is being treated as follow-up work, so this PR stays scoped to fixing the one metric.
